Expecta/ExpectaSupport.m

```diff
@@ -14,89 +15,69 @@ - (void)recordFailureWithDescription:(NSString *)description inFile:(NSString *)
       // @encode(EXPBasicBlock) returns @? as of clang 4.1.
       // This condition must occur before the test for id/class type,
       // otherwise blocks will be treated as vanilla objects.
-      id actual = va_arg(v, EXPBasicBlock);
+      id actual = *(EXPBasicBlock *)value;
       obj = [[actual copy] autorelease];
   } else if((strstr(type, @encode(id)) != NULL) || (strstr(type, @encode(Class)) != 0)) {
-    id actual = va_arg(v, id);
-    obj = actual;
+    obj = *(id *)value;
   } else if(strcmp(type, @encode(__typeof__(nil))) == 0) {
     obj = nil;
-  } else if(strstr(type, "ff}{") != NULL) { //TODO: of course this only works for a 2x2 e.g. CGRect
-    obj = [[[EXPFloatTuple alloc] initWithFloatValues:(float *)va_arg(v, float[4]) size:4] autorelease];
-  } else if(strstr(type, "=ff}") != NULL) {
-    obj = [[[EXPFloatTuple alloc] initWithFloatValues:(float *)va_arg(v, float[2]) size:2] autorelease];
-  } else if(strstr(type, "=fff}") != NULL) {
-    obj = [[[EXPFloatTuple alloc] initWithFloatValues:(float *)va_arg(v, float[3]) size:3] autorelease];
-  } else if(strstr(type, "=ffff}") != NULL) {
-    obj = [[[EXPFloatTuple alloc] initWithFloatValues:(float *)va_arg(v, float[4]) size:4] autorelease];
-  } else if(strstr(type, "dd}{") != NULL) { //TODO: same here
-    obj = [[[EXPDoubleTuple alloc] initWithDoubleValues:(double *)va_arg(v, double[4]) size:4] autorelease];
-  } else if(strstr(type, "=dd}") != NULL) {
-    obj = [[[EXPDoubleTuple alloc] initWithDoubleValues:(double *)va_arg(v, double[2]) size:2] autorelease];
-  } else if(strstr(type, "=dddd}") != NULL) {
-    obj = [[[EXPDoubleTuple alloc] initWithDoubleValues:(double *)va_arg(v, double[4]) size:4] autorelease];
+  } else if(strstr(type, "ff}{") != NULL || strstr(type, "=ff}") != NULL ||
+            strstr(type, "=fff}") != NULL || strstr(type, "=ffff}") != NULL) {
+    NSUInteger size;
+    NSGetSizeAndAlignment(type, &size, NULL);
+    obj = [[[EXPFloatTuple alloc] initWithFloatValues:(float *)value
+                                                 size:size / sizeof(float)] autorelease];
+  } else if(strstr(type, "dd}{") != NULL || strstr(type, "=dd}") != NULL ||
+            strstr(type, "=ddd}") != NULL || strstr(type, "=dddd}") != NULL) {
+    NSUInteger size;
+    NSGetSizeAndAlignment(type, &size, NULL);
+    obj = [[[EXPDoubleTuple alloc] initWithDoubleValues:(double *)value
+                                                   size:size / sizeof(double)] autorelease];
   } else if(type[0] == '{') {
     EXPUnsupportedObject *actual = [[[EXPUnsupportedObject alloc] initWithType:@"struct"] autorelease];
     obj = actual;
   } else if(type[0] == '(') {
     EXPUnsupportedObject *actual = [[[EXPUnsupportedObject alloc] initWithType:@"union"] autorelease];
     obj = actual;
   } else {
-    void *actual = va_arg(v, void *);
-    obj = (actual == NULL ? nil :[NSValue valueWithPointer:actual]);
+    void *actual = *(void **)value;
```

Comment:
can the entire cast be removed?

Reply:
Not really, since you need the dereference.